kstat.rs

```diff
@@ -0,0 +1,313 @@
+#![cfg(target_os = "solaris")]
+
+use std::os::raw::c_int;
+use std::os::raw::c_uint;
+use std::os::raw::c_char;
+use std::os::raw::c_uchar;
+use std::os::raw::c_void;
+use std::os::raw::c_long;
+use std::os::raw::c_ulong;
+use std::os::raw::c_longlong;
+use std::ptr::{null, null_mut};
+use std::ffi::CStr;
+use std::ffi::CString;
+
+type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+
+const KSTAT_STRLEN: usize = 31;
+
+const MODULE_CPU_INFO: &str = "cpu_info";
+
+const STAT_CLOCK_MHZ: &str = "clock_MHz";
+
+const MODULE_UNIX: &str = "unix";
+
+const NAME_SYSTEM_MISC: &str = "system_misc";
+const STAT_BOOT_TIME: &str = "boot_time";
+const STAT_NPROC: &str = "nproc";
+
+const NAME_SYSTEM_PAGES: &str = "system_pages";
+const STAT_FREEMEM: &str = "freemem";
+const STAT_PHYSMEM: &str = "physmem";
+
+
+#[repr(C)]
+struct Kstat {
+    ks_crtime: c_longlong,
+    ks_next: *mut Kstat,
+    ks_kid: c_uint,
+    ks_module: [c_char; KSTAT_STRLEN],
+    ks_resv: c_uchar,
+    ks_instance: c_int,
+    ks_name: [c_char; KSTAT_STRLEN],
+    ks_type: c_uchar,
+    ks_class: [c_char; KSTAT_STRLEN],
+    ks_flags: c_uchar,
+    ks_data: *mut c_void,
+    ks_ndata: c_uint,
+    ks_data_size: usize,
+    ks_snaptime: c_longlong,
+}
+
+impl Kstat {
+    unsafe fn name(&self) -> String {
+        CStr::from_ptr(self.ks_name.as_ptr()).to_str().unwrap().to_string()
+    }
+
+    unsafe fn module(&self) -> String {
+        CStr::from_ptr(self.ks_module.as_ptr()).to_str().unwrap().to_string()
+    }
+}
+
+#[repr(C)]
+struct KstatCtl {
+    kc_chain_id: c_uint,
+    kc_chain: *mut Kstat,
+    kc_kd: c_int,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+union KstatValue {
+    c: [c_char; 16],
+    l: c_long,
+    ul: c_ulong,
+    ui32: u32,
+}
+
+#[repr(C)]
+struct KstatNamed {
+    name: [c_char; KSTAT_STRLEN],
+    data_type: c_uchar,
+    value: KstatValue,
+}
+
+extern "C" {
+    fn kstat_open() -> *mut KstatCtl;
+    fn kstat_close(kc: *mut KstatCtl) -> c_int;
+    fn kstat_lookup(kc: *mut KstatCtl, module: *const c_char,
+        instance: c_int, name: *const c_char) -> *mut Kstat;
+    fn kstat_read(kc: *mut KstatCtl, ksp: *mut Kstat, buf: *mut c_void)
+        -> c_int;
+    fn kstat_data_lookup(ksp: *mut Kstat, name: *const c_char) -> *mut c_void;
+}
+
+/// Minimal wrapper around libkstat(3LIB) on illumos and Solaris systems.
+struct KstatWrapper {
```

Comment:
This would be easier to be confident about if the kstat stuff were in a separate module from its users. (Perhaps just a `mod kstat { ... }` right in this file.)

The reason: currently, any client can construct a `KstatWrapper` freely with bad pointers, without using `open`, since all the clients are right here in this module. That means that we can only be confident about the safety of operations on `KstatWrapper` by reading the clients, as well as the implementation.

Reply:
I have moved the `KstatWrapper` implementation into a child module.